packages/core/src/agent.ts

```diff
@@ -143,4 +149,31 @@ export class Agent<
     return await Promise.all(this.refs.map(ref => ref.applicable(ctx))).then(results => results.some(r => r));
   }
 
+  /**
+   * Returns metadata for the agent based on the input and context.
+   * Combines static metadata with dynamically generated metadata from metadataFn.
+   *
+   * @param input - The input for the agent.
+   * @param ctx - The context for the agent's operation.
+   * @returns The metadata for the agent.
+   */
+  async metadata<
+    TRuntimeContext extends TContext,
+    TRuntimeMetadata extends TMetadata,
+    TCoreContext extends Context<TRuntimeContext, TRuntimeMetadata>,
+  >(...[inputMaybe, contextMaybe]: OptionalParams<[TInput, TCoreContext]>): Promise<TMetadata> {
```

Comment:
This syntax may still make input and context be required by the dev, make it input:? TInput, context?: TCoreContext

Have two method signatures if possible - one no arg that doesn't return a promise - just TMetadata directly. Update all the places to behave in this way if possible